lib/Sema/CSSimplify.cpp

```diff
@@ -12530,7 +12555,8 @@ ConstraintSystem::simplifyKeyPathConstraint(
   if (definitelyFunctionType) {
     increaseScore(SK_FunctionConversion, locator);
     return SolutionKind::Solved;
-  } else if (!anyComponentsUnresolved ||
+  } else if ((!flags.contains(TMF_GenerateConstraints) &&
+              !anyComponentsUnresolved) ||
```

Comment:
@xedin one other thing for you to look at when you have a chance to check on this PR—I originally thought your recent changes to keypath simplification would remove the need for this, but we were still failing to properly convert the paths in the `testMinimalKeypaths()` test below.

Since those keypaths don't have anything to formally 'resolve' we were hitting this path during constraint generation before the proper disjunction/applicable function constraints had even been added to the system.

Also now probably makes sense to rename `anyComponentsUnresolved` since it's become extremely narrow, basically only used for code completion... I'll update to rename that soon.

Reply:
I think we need to figure out a better way here. Maybe, temporarily, it would be okay to just check

```
 if (Phase == ConstraintSystemPhase::ConstraintGeneration)
    return formUnsolved();
```

at the very start of this method.

But what @amritpan and I worked toward this summer (but couldn't quite land it yet) is moving key path "opening" into the `resolveKeyPath` and allowing it to happen only when there is either a contextual type or it's known that there wouldn't be one, this way we don't actually have to worry about eagerly resolving key path literals.

Reply:
Hmm how could we conclusively determine whether we'll be able to infer a type before we actually attempt solving? Right now, this is the only place we will 'default' to an actual `KeyPath` type if no other context is available, so if we move this to `resolveKeyPath` it seems like we'd need 100% accuracy, otherwise we risk either 1) binding to `KeyPath` too early when we should have inferred a function type somewhere along the way, or else 2) failing to bind to `KeyPath` because we were unsure, but then down the line failing to solve the system because we needed the `KeyPath` context to come from the literal...

Reply:
Also—what is the significance of using `flags` vs. `Phase` here? I guess `flags` could be set for constraint generation even if we are not actually walking an expression to set the `Phase`?

Reply:
> Hmm how could we conclusively determine whether we'll be able to infer a type before we actually attempt solving?

Only inference can tell us conclusively in certain conditions that's why resolveKeyPath is going to be called on that path (as in in `matchTypesBindTypeVar` just like we do for closures). We should only let `simplifyKeyPathConstraint` set the type if there is definitely contextual type we could use, once we have that we can remove this change to iterate over constraints I added as well.

> `Phase` is a safer bet because it identifies explicitly what phase constraint system is currently in, `TMF_GenerateConstraints` could technically be set randomly and not as narrow.

Reply:
Ah I was confusing `resolveKeyPath` and `resolveKeyPathExpr` from `PreCheckExpr` 🤦 This makes more sense!

Is there a WIP branch anywhere? I'll make the stopgap change in this PR but would be interested in taking a look that the proper fix next unless @amritpan you still had plans to tackle that?

Reply:
I don't think @amritpan opened a PR for the work she was doing because it wasn't complete but I guess it would be reasonable to open a draft PR?

Reply:
> unless @amritpan you still had plans to tackle that

[Here's a draft PR](https://github.com/apple/swift/pull/69286) of some of the `resolveKeyPathExpr` expansion that I worked on with @xedin, in case it helps.